File: ui/src/common/types/codingTask.ts
export type CodingTaskStatus =
  | "created"
  | "clarifying"
  | "spec_ready"
  | "planning"
  | "tasks_ready"
  | "generating"
  | "verifying"
  | "completed"
  | "failed"
  | "cancelled";

export type CodingTaskProfile = "hs_8bit_mcu" | "generic_coding";

export type SpecArtifactKind =
  | "requirement"
  | "spec"
  | "plan"
  | "tasks"
  | "checklist"
  | "acceptance"
  | "trace";

export interface CodingTask {
  id: string;
  title: string;
  workspace_path?: string | null;
  profile: CodingTaskProfile | string;
  target_chip?: string | null;
  project_type: string;
  status: CodingTaskStatus | string;
  selected_agent_id?: string | null;
  backend?: string | null;
  conversation_id?: number | null;
  selected_knowledge_scopes: string[];
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface CodingTaskCreateRequest {
  title: string;
  workspace_path?: string;
  profile?: CodingTaskProfile | string;
  target_chip?: string;
  project_type?: string;
  selected_agent_id?: string;
  backend?: string;
  conversation_id?: number;
  selected_knowledge_scopes?: string[];
  metadata?: Record<string, unknown>;
}

export interface SpecArtifact {
  id: string;
  coding_task_id: string;
  kind: SpecArtifactKind | string;
  title: string;
  content: string;
  format: string;
  status: string;
  trace_links: string[];
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface SpecArtifactUpsertRequest {
  task_id: string;
  kind: SpecArtifactKind;
  title: string;
  content: string;
  format?: string;
  status?: string;
  trace_links?: string[];
  metadata?: Record<string, unknown>;
}

File: crates/backend/nomifun-api-types/src/coding_task.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct CodingTaskCreateRequest {
    pub title: String,
    #[serde(default)]
    pub workspace_path: Option<String>,
    #[serde(default)]
    pub profile: Option<String>,
    #[serde(default)]
    pub target_chip: Option<String>,
    #[serde(default)]
    pub project_type: Option<String>,
    #[serde(default)]
    pub selected_agent_id: Option<String>,
    #[serde(default)]
    pub backend: Option<String>,
    #[serde(default)]
    pub conversation_id: Option<i64>,
    #[serde(default)]
    pub selected_knowledge_scopes: Vec<String>,
    #[serde(default)]
    pub metadata: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CodingTaskResponse {
    pub id: String,
    pub title: String,
    pub workspace_path: Option<String>,
    pub profile: String,
    pub target_chip: Option<String>,
    pub project_type: String,
    pub status: String,
    pub selected_agent_id: Option<String>,
    pub backend: Option<String>,
    pub conversation_id: Option<i64>,
    pub selected_knowledge_scopes: Vec<String>,
    pub metadata: serde_json::Value,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ListCodingTasksQuery {
    #[serde(default)]
    pub limit: Option<u32>,
    #[serde(default)]
    pub offset: Option<u32>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SpecArtifactUpsertRequest {
    pub title: String,
    #[serde(default)]
    pub content: String,
    #[serde(default)]
    pub format: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub trace_links: Vec<String>,
    #[serde(default)]
    pub metadata: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SpecArtifactResponse {
    pub id: String,
    pub coding_task_id: String,
    pub kind: String,
    pub title: String,
    pub content: String,
    pub format: String,
    pub status: String,
    pub trace_links: Vec<String>,
    pub metadata: serde_json::Value,
    pub created_at: String,
    pub updated_at: String,
}
